core/src/main/java/com/radixdlt/messaging/core/MessageDispatcher.java

```diff
@@ -101,6 +106,12 @@ class MessageDispatcher {
   private final PeerManager peerManager;
   private final Addressing addressing;
 
+  private final RateLimiter ttlExpiredLogRateLimiter = RateLimiter.create(0.05);
+  private final AtomicInteger ttlExpiredMessagesCountSinceLastLog = new AtomicInteger(0);
+
+  private final LRUCache<NodeId, Pair<RateLimiter, AtomicInteger>>
```

Comment:
It looks like a pair `(RateLimiter, AtomicInteger)` actually represents a "rate-limited logger":
1. I vaguely remember that some generic solutions exist for that (for each major logging framework) and it would be cool to use one of them instead.
2. But I also see that you have some more specific needs (e.g. "apply rate-limit per receiver"), so maybe we need to implement it ourselves - but even in such case, I would strongly advice to extract a `class LogLimiter` and then have `private LogLimiter ttlExpired; private LRUCache<NodeId, LogLimiter> sendErrors;`

Reply:
yeah, there are some configurable filters but I haven't found anything that matched my requirements :)
I extracted `RejectionCountingRateLimiter` (it doesn't really need to be logger-specific).